File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/InstructionBox.java
package org.hyperledger.iroha.android.model;

import java.util.Base64;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.model.instructions.InstructionKind;

/**
 * Typed representation of an instruction scheduled for execution within a transaction.
 *
 * <p>The container wraps strongly typed instruction payloads where available (for example, register
 * domain/account/asset definition builders) and can wrap opaque key/value payloads for local use.
 * Transaction encoding requires wire-framed instruction payloads so Norito parity is preserved
 * whenever raw instruction bytes are available.
 */
public final class InstructionBox {

  private static final String ARG_WIRE_NAME = "wire_name";
  private static final String ARG_PAYLOAD_BASE64 = "payload_base64";

  private final InstructionPayload payload;

  private InstructionBox(final InstructionPayload payload) {
    this.payload = Objects.requireNonNull(payload, "payload");
  }

  /** Instruction display name (matches `InstructionType` tag). */
  public String name() {
    if (payload instanceof WirePayload wire) {
      return wire.wireName();
    }
    return kind().displayName();
  }

  public InstructionKind kind() {
    return payload.kind();
  }

  public InstructionPayload payload() {
    return payload;
  }

  /**
   * Returns Norito-ready arguments describing the instruction. The returned map is immutable and
   * safe to cache by callers.
   */
  public Map<String, String> arguments() {
    return payload.toArguments();
  }

  public Builder toBuilder() {
    return builder().setKind(kind()).setArguments(arguments());
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof InstructionBox other)) {
      return false;
    }
    return kind() == other.kind() && arguments().equals(other.arguments());
  }

  @Override
  public int hashCode() {
    return Objects.hash(kind(), arguments());
  }

  public static InstructionBox of(final InstructionPayload payload) {
    return new InstructionBox(payload);
  }

  /**
   * Builds an {@link InstructionBox} from a canonical wire identifier and its Norito-framed payload.
   *
   * <p>The payload must include the Norito header with a valid checksum.
   */
  public static InstructionBox fromWirePayload(final String wireName, final byte[] payloadBytes) {
    return new InstructionBox(new WireInstructionPayload(wireName, payloadBytes));
  }

  /**
   * Builds an {@link InstructionBox} from Norito decoded components.
   *
   * <p>The arguments must include {@code wire_name} and {@code payload_base64} fields carrying a
   * Norito-framed instruction payload.
   */
  public static InstructionBox fromNorito(
      final InstructionKind kind, final Map<String, String> arguments) {
    return new InstructionBox(resolvePayload(kind, arguments));
  }

  public static Builder builder() {
    return new Builder();
  }

  private static InstructionPayload resolvePayload(
      final InstructionKind kind, final Map<String, String> arguments) {
    Objects.requireNonNull(kind, "kind");
    Objects.requireNonNull(arguments, "arguments");
    final String wireName = arguments.get(ARG_WIRE_NAME);
    final String payloadBase64 = arguments.get(ARG_PAYLOAD_BASE64);
    if (wireName == null && payloadBase64 == null) {
      throw new IllegalArgumentException(
          "instruction arguments must include wire_name and payload_base64 fields");
    }
    if (wireName == null || payloadBase64 == null) {
      throw new IllegalArgumentException(
          "wire payload requires both wire_name and payload_base64 fields");
    }
    final byte[] payloadBytes;
    try {
      payloadBytes = Base64.getDecoder().decode(payloadBase64);
    } catch (final IllegalArgumentException ex) {
      throw new IllegalArgumentException("wire payload base64 is invalid", ex);
    }
    return new WireInstructionPayload(wireName, payloadBytes);
  }

  private static InstructionKind wireKindForName(final String wireName) {
    if (wireName == null) {
      return InstructionKind.CUSTOM;
    }
    final String normalized = wireName.toLowerCase(Locale.ROOT);
    if (normalized.startsWith("iroha.register")) {
      return InstructionKind.REGISTER;
    }
    if (normalized.startsWith("iroha.unregister")) {
      return InstructionKind.UNREGISTER;
    }
    if (normalized.startsWith("iroha.transfer")) {
      return InstructionKind.TRANSFER;
    }
    if (normalized.startsWith("iroha.mint")) {
      return InstructionKind.MINT;
    }
    if (normalized.startsWith("iroha.burn")) {
      return InstructionKind.BURN;
    }
    if (normalized.startsWith("iroha.grant")) {
      return InstructionKind.GRANT;
    }
    if (normalized.startsWith("iroha.revoke")) {
      return InstructionKind.REVOKE;
    }
    if (normalized.startsWith("iroha.set_key_value")) {
      return InstructionKind.SET_KEY_VALUE;
    }
    if (normalized.startsWith("iroha.remove_key_value")) {
      return InstructionKind.REMOVE_KEY_VALUE;
    }
    if (normalized.startsWith("iroha.set_parameter")) {
      return InstructionKind.SET_PARAMETER;
    }
    if (normalized.startsWith("iroha.execute_trigger")) {
      return InstructionKind.EXECUTE_TRIGGER;
    }
    if (normalized.startsWith("iroha.log")) {
      return InstructionKind.LOG;
    }
    if (normalized.startsWith("iroha.runtime_upgrade") || normalized.startsWith("iroha.upgrade")) {
      return InstructionKind.UPGRADE;
    }
    return InstructionKind.CUSTOM;
  }

  public interface InstructionPayload {
    InstructionKind kind();

    Map<String, String> toArguments();
  }

  public interface WirePayload extends InstructionPayload {
    String wireName();

    byte[] payloadBytes();
  }

  private static final class WireInstructionPayload implements WirePayload {
    private final InstructionKind kind;
    private final String wireName;
    private final byte[] payloadBytes;
    private final Map<String, String> arguments;

    private WireInstructionPayload(final String wireName, final byte[] payloadBytes) {
      if (wireName == null || wireName.isBlank()) {
        throw new IllegalArgumentException("wireName must not be blank");
      }
      if (payloadBytes == null || payloadBytes.length == 0) {
        throw new IllegalArgumentException("payloadBytes must not be empty");
      }
      this.wireName = wireName;
      this.payloadBytes = payloadBytes.clone();
      this.kind = wireKindForName(wireName);
      final LinkedHashMap<String, String> args = new LinkedHashMap<>();
      args.put(ARG_WIRE_NAME, wireName);
      args.put(ARG_PAYLOAD_BASE64, Base64.getEncoder().encodeToString(this.payloadBytes));
      this.arguments = Collections.unmodifiableMap(args);
    }

    @Override
    public InstructionKind kind() {
      return kind;
    }

    @Override
    public Map<String, String> toArguments() {
      return arguments;
    }

    @Override
    public String wireName() {
      return wireName;
    }

    @Override
    public byte[] payloadBytes() {
      return payloadBytes.clone();
    }
  }

  public static final class Builder {
    private InstructionKind kind = InstructionKind.CUSTOM;
    private final Map<String, String> arguments = new LinkedHashMap<>();

    public Builder setKind(final InstructionKind kind) {
      this.kind = Objects.requireNonNull(kind, "kind");
      return this;
    }

    public Builder setName(final String name) {
      Objects.requireNonNull(name, "name");
      try {
        this.kind = InstructionKind.fromDisplayName(name);
      } catch (final IllegalArgumentException ignored) {
        this.kind = InstructionKind.CUSTOM;
      }
      arguments.putIfAbsent("action", name);
      return this;
    }

    public Builder putArgument(final String key, final String value) {
      arguments.put(Objects.requireNonNull(key, "key"), Objects.requireNonNull(value, "value"));
      return this;
    }

    public Builder setArguments(final Map<String, String> arguments) {
      this.arguments.clear();
      if (arguments != null) {
        arguments.forEach(this::putArgument);
      }
      return this;
    }

    public InstructionBox build() {
      return InstructionBox.fromNorito(kind, arguments);
    }
  }
}

File: .taira-source-c80bf4d85a97/kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/InstructionBox.kt
@file:OptIn(ExperimentalEncodingApi::class)

package org.hyperledger.iroha.sdk.core.model

import org.hyperledger.iroha.sdk.core.model.instructions.InstructionKind
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

private const val ARG_WIRE_NAME = "wire_name"
private const val ARG_PAYLOAD_BASE64 = "payload_base64"

/**
 * Typed representation of an instruction scheduled for execution within a transaction.
 *
 * The container wraps strongly typed instruction payloads where available (for example, register
 * domain/account/asset definition builders) and can wrap opaque key/value payloads for local use.
 * Transaction encoding requires wire-framed instruction payloads so Norito parity is preserved
 * whenever raw instruction bytes are available.
 */
class InstructionBox private constructor(val payload: InstructionPayload) {

    /** Instruction display name (matches `InstructionType` tag). */
    val name: String
        get() = when (payload) {
            is WirePayload -> payload.wireName
            else -> kind.displayName
        }

    val kind: InstructionKind get() = payload.kind

    /**
     * Returns Norito-ready arguments describing the instruction. The returned map is immutable and
     * safe to cache by callers.
     */
    val arguments: Map<String, String> get() = payload.arguments

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is InstructionBox) return false
        return kind == other.kind && arguments == other.arguments
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + arguments.hashCode()
        return result
    }

    companion object {
        @JvmStatic
        fun of(payload: InstructionPayload): InstructionBox = InstructionBox(payload)

        /**
         * Builds an `InstructionBox` from a canonical wire identifier and its Norito-framed payload.
         *
         * The payload must include the Norito header with a valid checksum.
         */
        @JvmStatic
        fun fromWirePayload(wireName: String, payloadBytes: ByteArray): InstructionBox =
            InstructionBox(WireInstructionPayload(wireName, payloadBytes))

        /**
         * Builds an `InstructionBox` from Norito decoded components.
         *
         * The arguments must include `wire_name` and `payload_base64` fields carrying a
         * Norito-framed instruction payload.
         */
        @JvmStatic
        fun fromNorito(arguments: Map<String, String>): InstructionBox =
            InstructionBox(resolvePayload(arguments))

        private fun resolvePayload(arguments: Map<String, String>): InstructionPayload {
            val wireName = arguments[ARG_WIRE_NAME]
            val payloadBase64 = arguments[ARG_PAYLOAD_BASE64]
            require(wireName != null || payloadBase64 != null) {
                "instruction arguments must include wire_name and payload_base64 fields"
            }
            require(wireName != null && payloadBase64 != null) {
                "wire payload requires both wire_name and payload_base64 fields"
            }
            val payloadBytes = try {
                Base64.decode(payloadBase64)
            } catch (ex: IllegalArgumentException) {
                throw IllegalArgumentException("wire payload base64 is invalid", ex)
            }
            return WireInstructionPayload(wireName, payloadBytes)
        }
    }

    private class WireInstructionPayload(
        override val wireName: String,
        payloadBytes: ByteArray,
    ) : WirePayload {

        init {
            require(wireName.isNotBlank()) { "wireName must not be blank" }
            require(payloadBytes.isNotEmpty()) { "payloadBytes must not be empty" }
        }

        private val _payloadBytes: ByteArray = payloadBytes.copyOf()

        override val kind: InstructionKind = wireKindForName(wireName)

        override val arguments: Map<String, String> = mapOf(
            ARG_WIRE_NAME to wireName,
            ARG_PAYLOAD_BASE64 to Base64.encode(_payloadBytes),
        )

        override val payloadBytes: ByteArray get() = _payloadBytes.copyOf()
    }
}

private fun wireKindForName(wireName: String): InstructionKind {
    val normalized = wireName.lowercase()
    return when {
        normalized.startsWith("iroha.register") -> InstructionKind.REGISTER
        normalized.startsWith("iroha.unregister") -> InstructionKind.UNREGISTER
        normalized.startsWith("iroha.transfer") -> InstructionKind.TRANSFER
        normalized.startsWith("iroha.mint") -> InstructionKind.MINT
        normalized.startsWith("iroha.burn") -> InstructionKind.BURN
        normalized.startsWith("iroha.grant") -> InstructionKind.GRANT
        normalized.startsWith("iroha.revoke") -> InstructionKind.REVOKE
        normalized.startsWith("iroha.set_key_value") -> InstructionKind.SET_KEY_VALUE
        normalized.startsWith("iroha.remove_key_value") -> InstructionKind.REMOVE_KEY_VALUE
        normalized.startsWith("iroha.set_parameter") -> InstructionKind.SET_PARAMETER
        normalized.startsWith("iroha.execute_trigger") -> InstructionKind.EXECUTE_TRIGGER
        normalized.startsWith("iroha.log") -> InstructionKind.LOG
        normalized.startsWith("iroha.runtime_upgrade") || normalized.startsWith("iroha.upgrade") -> InstructionKind.UPGRADE
        else -> InstructionKind.CUSTOM
    }
}
